src/domain/community/community/community.service.ts

```diff
@@ -0,0 +1,317 @@
+        'Community not set',
+        LogContext.COMMUNITY
+      );
+    community.parentCommunity = parentCommunity;
+    await this.communityRepository.save(community);
+    return community;
+  }
+
+  async addMember(userID: number, communityID: number): Promise<IUserGroup> {
+    const community = await this.getCommunityOrFail(communityID, {
+      relations: ['groups', 'parentCommunity'],
+    });
+
+    // If the parent community is set, then check if the user is also a member there
+    if (community.parentCommunity) {
+      const isParentMember = await this.isMember(
+        userID,
+        community.parentCommunity.id
+      );
+      if (!isParentMember)
+        throw new ValidationException(
+          `User (${userID}) is not a member of parent community: ${community.parentCommunity.name}`,
+          LogContext.CHALLENGES
+        );
+    }
+
+    // Try to find the user + group
+    const user = await this.userService.getUserByIdOrFail(userID);
+
+    // Get the members group
+    const membersGroup = await this.getMembersGroupOrFail(community);
+    await this.userGroupService.addUserToGroup(user, membersGroup);
+
+    return membersGroup;
+  }
+
+  async isMember(userID: number, communityID: number): Promise<boolean> {
+    const community = await this.getCommunityOrFail(communityID, {
+      relations: ['groups'],
+    });
+
+    const members = await this.getMembersOrFail(community);
+    const user = members.find(user => user.id == userID);
+    if (user) return true;
+    return false;
+  }
+
+  async getMembersGroupOrFail(community: ICommunity): Promise<IUserGroup> {
+    const group = await this.userGroupService.getGroupByName(
+      community,
+      AuthorizationRoles.Members
+    );
+    if (!group)
+      throw new RelationshipNotFoundException(
+        `Unable to locate members group on Community: ${community.id}`,
+        LogContext.COMMUNITY
+      );
+    return group;
+  }
+
+  async getMembersOrFail(community: ICommunity): Promise<IUser[]> {
+    const membersGroup = await this.getMembersGroupOrFail(community);
+    const members = membersGroup.members;
+    if (!members)
+      throw new GroupNotInitializedException(
+        `Members group not initialised in Community: ${community.id}`,
+        LogContext.COMMUNITY
+      );
+
+    return members;
+  }
+
+  async getCommunities(ecoverseId: number): Promise<Community[]> {
+    const communites = await this.communityRepository.find({
+      where: { ecoverse: { id: ecoverseId } },
+    });
+    return communites || [];
+  }
+
+  async createApplication(
+    id: number,
+    applicationData: ApplicationInput
+  ): Promise<Application> {
+    const community = (await this.getCommunityOrFail(id, {
+      relations: ['applications'],
+    })) as Community;
+
+    const existingApplication = community.applications?.find(
+      x => x.user.id === applicationData.userId
+    );
+
+    if (existingApplication) {
+      throw new ApolloError(
```

Comment:
Wrap this in a custom exception. We don't use ApolloErrors directly - we have some custom metadata in our exceptions.

Reply:
Fixed in https://github.com/cherrytwist/Server/pull/821/commits/65fe1f0bdf1729691b2b15de0e04b3e9e6a36c13